### PR TITLE
[Kraken] private channel

### PR DESCRIPTION
**Pull request by mdvx**
A first functional pass at Kraken private channel
Known Issues
* getOrderChanges & getUserTrades ignore currencyPair parameter (and subscribe to all)
** now does its best to filter
* does not unsubscribe from channels (ownTrades & openOrders)

UnKnown Issues
* possibily some

### CODE REVIEW

**Comment by badgerwithagun** [comment]
Is this ready for review, @mdvx ?

**Comment by mdvx** [comment]
Yes, lets do a first review - please

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 1]
```diff
@@ -18,47 +17,72 @@
...
+    private KrakenStreamingTradeService streamingTradeService;
+
     private final SynchronizedValueFactory<Long> nonceFactory = new CurrentTimeNonceFactory();
 
     public KrakenStreamingExchange() {
-        this.streamingService = new KrakenStreamingService(API_URI);
+    }
+
+    private static String pickUri(boolean isPrivate, Boolean useSandbox) {
+        if (useSandbox != null && useSandbox)
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Is the beta environment really a "sandbox"?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 2]
```diff
@@ -18,47 +17,72 @@
...
+            return API_BETA_URI;
+        else
+            return isPrivate ? API_AUTH_URI : API_URI;
     }
 
     @Override
     protected void initServices() {
         super.initServices();
-        streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+        Boolean useSanbox = (Boolean)exchangeSpecification.getExchangeSpecificParametersItem(USE_SANDBOX);
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: `useSandbox` (spelling).

Also, if you use `MoreObjects.firstNonNull((Boolean)exchangeSpecification.getExchangeSpecificParametersItem(USE_SANDBOX), false)` you can avoid downstream null checks

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 3]
```diff
@@ -18,47 +17,72 @@
...
+
+        this.streamingService = new KrakenStreamingService(false, pickUri(false,useSanbox));
+        this.streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+
+        if (exchangeSpecification.getApiKey() != null) {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Use `StringUtils.isNotEmpty`; when substituting environment variables into application config, it's hard to force a configuration value to null.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 4]
```diff
@@ -18,47 +17,72 @@
...
     @Override
     public boolean isAlive() {
-        return streamingService.isSocketOpen();
+        return streamingService.isSocketOpen() && (privateStreamingService == null || privateStreamingService.isSocketOpen());
     }
     
     @Override
     public Observable<Object> connectionSuccess() {
-        return streamingService.subscribeConnectionSuccess();
+        return streamingService.subscribeConnectionSuccess().mergeWith(privateStreamingService.subscribeConnectionSuccess());
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Hmmm. It seems a little unexpected that you'll get two connection successes. I wonder what people use this observable for?  With this possibility, you can't rely on it for state on its own; you need to check `isAlive()` as well, which should probably documented if we think this is the right behaviour.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 5]
```diff
@@ -18,47 +17,72 @@
...
     }
     
     @Override
     public Observable<Throwable> reconnectFailure() {
-        return streamingService.subscribeReconnectFailure();
+        return streamingService.subscribeReconnectFailure().mergeWith(privateStreamingService.subscribeReconnectFailure());
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Same applies here, possibly

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java [hunk 6]
```diff
@@ -27,18 +27,20 @@
  * @author makarid, pchertalev
  */
 public class KrakenStreamingService extends JsonNettyStreamingService {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java]: Looking at this code, there seems to be so many `if.. else` sections, I wonder if it might be more clearly implemented as two separate classes, one private and one public. Thoughts?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            token = rawKrakenAcctService.getKrakenWebsocketToken();
+            tokenExpires = System.currentTimeMillis() + (token.getExpiresInSeconds() * 900L); // 900L instead of a 1000L for 90%
+        }
+        return token;
+    }
+
+    @Override
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+
+        return Observable.create( source -> {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Why the wrapper observable? Looks like you can just return the result of `subscribeChannel`. Something like this (probably doesn't work, but certainly something like this):
```
return streamingService.subscribeChannel(getChannelName(KrakenSubscriptionName.openOrders), renewToken().getToken())
    .filter(JsonNode::isArray)
    .filter(Objects::nonNull)
    .map(jsonNode -> jsonNode.get(0))
    .map(jsonNode -> StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
    .flatMapIterable(this::adaptKrakenOrders);
```

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                                adaptKrakenOrders(list)
+                                        .forEach(source::onNext));
+
+            } catch (Throwable e) {
+                source.onError(e);
+            }
+        });
+    }
+
+    private List<Order> adaptKrakenOrders(KrakenDtoOrderHolder[] dtoList) {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: If you changed this method to return `Iterable<Order>` you could do the conversion using an Guava `FluentIterable` and avoid creating a temporary list. Minor suggestion, not required to pass review.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                Order.OrderType side = descr == null ? null : KrakenAdapters.adaptOrderType(KrakenType.fromString(descr.type));
+
+                Date time = dto.opentm == null ? null : new Date((long) (dto.opentm * 1000L));
+                String ordertype = (descr == null || descr.ordertype == null) ? null : descr.ordertype;
+
+                Set<Order.IOrderFlags> flags = adaptFlags(dto.oflags);
+
+                if ("limit".equals(ordertype))
+                    result.add(
+                            new LimitOrder.Builder(side, pair)
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Most of these blocks are duplicated.  `LimitOrder.Builder` extends `Order.Builder` so you can create the builder in this `if` block, filling in the order-type specific bits, and then fill in all the common parts afterwards.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+        for (String flag : arr) {
+            flags.add(KrakenOrderFlags.fromString(flag));
+        }
+        return flags;
+    }
+
+    @Override
+    public Observable<UserTrade> getUserTrades(CurrencyPair currencyPair, Object... args) {
+        return Observable.create( source -> {
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: Ditto - seems unnecessary to create a new `Observable` here

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 1 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Not clear: but I changed the flag to Use_Beta
https://support.kraken.com/hc/en-us/articles/360000919926-Does-Kraken-offer-a-Test-API-or-Sandbox-Mode-

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java (hunk 6 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java]: It looks the next version  (beta-ws)  will merge the two back together (private and public), so might be worth waiting before creating new class for private?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java (hunk 8 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: no objection, it does not break anything.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java (hunk 9 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: fixed

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java (hunk 7 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: that works, and I added the missing currencyPair filter

Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java]: [Kraken] review fix #3
This is not great, we get updates without CurrencyPair, so we have to let those thru,
just in case they are for our subscription  (would have to maintain order cache otherwise)

                    .filter(order -> currencyPair == null || order.getCurrencyPair() == null || order.getCurrencyPair().compareTo(currencyPair) == 0);

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 4 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Not sure what to do about this.  Wait for both, then emit one.  kind of like debounce.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 5 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: for reconnectFailure, I think it is OK to get two errors

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 4 above)
Comment (badgerwithagun) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: I suppose that's probably the most predictable option, yes!

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 5 above)
Comment (badgerwithagun) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Ooo - what you could do here is have `streamingService` and `privateStreamingService` return different exception types?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java (hunk 6 above)
Comment (badgerwithagun) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java]: Fair enough.

File: xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java [hunk 11]
```diff
@@ -19,6 +19,8 @@
 public class CoinbaseProStreamingExchange extends CoinbaseProExchange implements StreamingExchange {
     private static final String API_URI = "wss://ws-feed.pro.coinbase.com";
     private static final String SANDBOX_API_URI = "wss://ws-feed-public.sandbox.pro.coinbase.com";
+    private static final String PRIME_API_URI = "wss://ws-feed.prime.coinbase.com";
```
Comment (badgerwithagun) [inline xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java]: Has this got mixed up with #476 ?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 12]
```diff
@@ -18,46 +20,77 @@
...
         super.initServices();
-        streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+        Boolean useBeta = MoreObjects.firstNonNull((Boolean)exchangeSpecification.getExchangeSpecificParametersItem(USE_BETA), Boolean.FALSE);
+
+        this.streamingService = new KrakenStreamingService(false, pickUri(false,useBeta));
+        this.streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+
+        if (StringUtils.isNotEmpty(exchangeSpecification.getApiKey())) {
+            this.privateStreamingService = new KrakenStreamingService(true, pickUri(true,useBeta));
+//            this.futuresStreamingService = new KrakenStreamingService(true, API_FUTURES_URI);
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Commented out code. Presumably this also means there's a compiler warning on the `API_FUTURES_API` constant.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 5 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: I like that, let me muse

File: xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java (hunk 11 above)
Comment (mdvx) [reply xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java]: yup

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 12 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: fixed

File: xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java (hunk 11 above)
Comment (mdvx) [reply xchange-coinbasepro/src/main/java/info/bitrich/xchangestream/coinbasepro/CoinbaseProStreamingExchange.java]: fixed in commit d80358f69ab5f37c1d6d5d912e8d2b0ebb23e0d6

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 5 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: 558e6f61ed4922f3fea4625241a71f6a8fefac0d

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 4 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: 558e6f61ed4922f3fea4625241a71f6a8fefac0d

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 13]
```diff
@@ -18,37 +22,61 @@
...
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Best to remove this commented out code. Code related to this is also going to be needed in other places (`connect`/`disconnect` etc) so it's less confusing to leave it out and start a different PR to add futures.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java [hunk 14]
```diff
@@ -111,36 +136,98 @@ protected String getChannelNameFromMessage(JsonNode message) throws IOException
...
-        String pair = channelData[1];
         KrakenSubscriptionName subscriptionName = KrakenSubscriptionName.valueOf(channelData[0]);
 
-        int reqID = Math.abs(UUID.randomUUID().hashCode());
-        subscriptionRequestMap.put(reqID, channelName);
-        KrakenSubscriptionMessage subscriptionMessage = new KrakenSubscriptionMessage(reqID, KrakenEventType.unsubscribe,
-                Collections.singletonList(pair), new KrakenSubscriptionConfig(subscriptionName));
-        return objectMapper.writeValueAsString(subscriptionMessage);
+        if (isPrivate) {
+            String token = null; // TODO: We probably need to pass the token to unsubscribe (but we dont have args[0])
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java]: TODO

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 15]
```diff
@@ -80,17 +81,11 @@ public boolean isAlive() {
     
     @Override
     public Observable<Object> connectionSuccess() {
-        if (privateStreamingService != null)
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Why was this removed?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java [hunk 16]
```diff
@@ -80,17 +81,11 @@ public boolean isAlive() {
...
-            return streamingService.subscribeConnectionSuccess().mergeWith(privateStreamingService.subscribeConnectionSuccess());
-
         return streamingService.subscribeConnectionSuccess();
     }
     
     @Override
     public Observable<Throwable> reconnectFailure() {
-        if (privateStreamingService != null)
```
Comment (badgerwithagun) [inline xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: Why was this removed?

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 13 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: done

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 16 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: because I cannot resolve this, so going for the simplest option of just monitoring the public channel.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java (hunk 15 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java]: because I cannot resolve this, so going for the simplest option of just monitoring the public channel.

File: xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java (hunk 14 above)
Comment (mdvx) [reply xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingService.java]: removed
